### PR TITLE
PER-8985 feat: toHaveScreenshot drop-in — route existing Playwright assertions through Percy

### PR DESCRIPTION
**Pull request by Shivanshu-07**
## What

Ships the `toHaveScreenshot()` drop-in inside `@percy/playwright` (`require('@percy/playwright/dropin')` — one config line, zero test changes). Every existing Playwright visual assertion routes through Percy:

- **Web project token** → serialized-DOM web snapshot (same capture as `percySnapshot`: readiness gate, responsive capture, CORS iframes; Locator subjects become scoped snapshots), rendered server-side.
- **App project token** → the captured PNG uploads straight through the comparison ingest (tag + tile) — no render flow, matching App Percy's existing ingestion.
- Any other token (automate/generic) → a clear wrong-token configuration error.

Assertions always pass locally (verdict moves to Percy review); Percy being down falls back to native `toHaveScreenshot` for the whole run; optional sync mode and a CI gate reporter are included.

## Baseline seeding (CLI-driven)

- `dropin/baseline/provider.js` — the `"@percy/cli".baselineProvider` hook `percy exec` discovers: resolves the Playwright config outside a test run (config-capture reporter under `playwright test --list`, so TS configs work) and maps committed `*-snapshots` PNGs onto Percy identity.
- `commands/setup-baseline.mjs` — `percy playwright:setup-baseline`, the explicit (re-)baseline command for established projects (storybook-style command registration).

## Depends on

- percy/cli — client build attributes + exec seeding orchestration (companion PR)
- percy/percy-api#6178 + companion API PR — baseline auto-approval

## Testing

126 specs green (dispatch, identity, provider, config-resolution units + real-runner integration); E2E verified locally against percy-api and on staging (web + app projects, all four flows).

## Automate project support (f8b7e18)

`automate` is now a supported drop-in project type: the matcher dispatches to the SDK's own `percyScreenshot()` Percy-on-Automate flow via a new `dropin/automate.js` seam — the capture happens on the remote BrowserStack browser through the Automate session. Details:

- An automate token on a **local** browser raises a clear per-assertion configuration error (not latched, so hybrid local+BrowserStack suites keep working for the remote projects).
- Locator subjects degrade to full-page with a one-time warning (remote capture has no element scoping); local-only pixel options (`mask`, `clip`, ...) are ignored with a one-time warning; `fullPage` and `sync` map through.
- Baseline seeding stays web/app-only: `playwright:setup-baseline` on an automate token exits with guidance ("Automate projects build their baseline from the first run on BrowserStack; no seeding is needed").

Companion changes: percy/cli#2335 (source tagging + seeding skip), percy/percy-api#6630 (server-side guards).
🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by Shivanshu-07** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/percy-playwright/pull/640  •  **Head:** ed2682a  •  **Reviewers:** fallback inline checklist (deep-review agent)

## Summary
Adds the `toHaveScreenshot()` drop-in: matcher override with web/app dispatch (percySnapshot DOM vs comparison ingest), always-pass posture, native fallback, baseline provider + config-capture reporter, setup-baseline command, and opt-in CI gate reporter.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | |
| High | Security | Authentication/authorization checks present | N/A | SDK — token handled by CLI/env |
| High | Security | Input validation and sanitization | Pass | token-type validation; sanitizeForFilePath parity |
| High | Security | No IDOR — resource ownership validated | N/A | |
| High | Security | No SQL injection (parameterized queries) | N/A | |
| High | Correctness | Logic is correct, handles edge cases | Pass | 3 High + 5 Medium findings fixed in ed2682a (gate exit status, exec deadlock, browserFamily parity, validation latch, informational-mode gate, trimLongString parity, CLI-version guard) |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | D3 always-pass swallowing is by design and documented |
| High | Correctness | No race conditions or concurrency issues | Pass | temp-file handoff uses pid+randomBytes; spawn is synchronous |
| Medium | Testing | New code has corresponding tests | Pass | 131 tests + new reporter/identity/latch coverage added |
| Medium | Testing | Error paths and edge cases tested | Pass | gate red-run, pending bow-out, isFailing matrix, .not, latch |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | full suite green (1 known shared-server flake passes on rerun) |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | |
| Medium | Performance | Long-running tasks use background jobs | N/A | |
| Medium | Quality | Follows existing codebase patterns | Pass | |
| Medium | Quality | Changes are focused (single concern) | Pass | |
| Low | Quality | Meaningful names, no dead code | Pass | dead code removed in ed2682a (base-branch.js, runNativeMatcher, isWriteOnlyToken) |
| Low | Quality | Comments explain why, not what | Pass | |
| Low | Quality | No unnecessary dependencies added | Pass | only @percy/sdk-utils floor |

## Findings

All review findings were fixed on this PR before this report (commit ed2682a):
- **High** `dropin/reporter.js`: gate could never red CI (Playwright discards `process.exitCode`) → onEnd now returns `{ status: 'failed' }`.
- **High** `dropin/reporter.js`: gate deadlocked ~10min under `percy exec` (build finalizes after the test process exits) → pending-grace probe + `build:wait` guidance.
- **High** `dropin/identity.js`: head browserFamily used the project display name, baselines used `use.browserName` → identity unified on browserName.
- **Medium**: config-validation latch (error now re-thrown on every assertion), informational gate never reds on errored builds, trimLongString parity for >100-char stems, setup-baseline CLI-version guard, reporter/sync/fallback test coverage.
- **Low**: Windows-safe playwright spawn, snapshotDir wiring, `.not` always-pass handling, heightless-tile skip, dead code removal, stale types refresh, engines/peer floors.

---

**Verdict: PASS**

File: commands/setup-baseline.mjs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    skipDiscovery: true
+  }
+}, async function * ({ percy, log, exit }) {
+  if (!percy) exit(1, 'Percy is disabled');
+
+  // Web projects seed rendered snapshots; app projects seed raw comparison uploads (no render
+  // flow — like App Percy). Anything else is a wrong token.
+  let projectType = percy.client.tokenType();
+  if (projectType !== 'web' && projectType !== 'app') {
+    exit(1, 'playwright:setup-baseline requires a Percy web or app project token — ' +
```
Comment (rishigupta1599) [inline commands/setup-baseline.mjs]: this is duplicate code, check if can be shifted to utils?

File: dropin/baseline/config-reporter.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+          viewport: p.use && p.use.viewport
+        },
+        snapshotPathTemplate: p.snapshotPathTemplate,
+        expect: screenshotTpl(p)
+          ? { toHaveScreenshot: { pathTemplate: screenshotTpl(p) } }
+          : undefined
+      }))
+    };
+
+    fs.writeFileSync(out, JSON.stringify(shape));
```
Comment (rishigupta1599) [inline dropin/baseline/config-reporter.js:44]: we are writing to file, are we using it somewhere or uploading?

File: dropin/png.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+// the newer /percy/screenshot endpoint does — so the SDK must send accurate dims itself. The PNG
+// bytes are the source of truth: IHDR is always the first chunk, width/height big-endian at
+// offsets 16/20.
+function pngDimensions(buf) {
+  if (!Buffer.isBuffer(buf) || buf.length < 24) return null;
+  // \x89PNG\r\n\x1a\n signature + "IHDR" chunk type at offset 12.
+  if (buf.readUInt32BE(12) !== 0x49484452) return null;
+  const width = buf.readUInt32BE(16);
+  const height = buf.readUInt32BE(20);
+  if (!width || !height) return null;
```
Comment (rishigupta1599) [inline dropin/png.js:15]: don't we have any standard package for png operations

File: dropin/version-info.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+// Shared client/environment-info strings sent on every Percy post (clientInfo /
+// environmentInfo). Computed once and reused across the drop-in modules so they don't each
+// recompute the identical pkg-name/version + playwright-version lookup.
+const pkg = require('../package.json');
```
Comment (rishigupta1599) [inline dropin/version-info.js]: this should not be inside droping it can be in general use right?
how are we capturing right now for user-agent creation?

File: dropin/sync.js [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+async function preflightTokenScope({ token = process.env.PERCY_TOKEN, probe, buildId, client } = {}) {
+  if (probe) {
+    try {
+      const readable = await probe(token);
+      return readable
+        ? { ok: true }
+        : { ok: false, reason: 'token cannot read comparison results' };
+    } catch (err) {
+      return { ok: false, reason: `token scope check failed — ${err.message}` };
+    }
```
Comment (rishigupta1599) [inline dropin/sync.js]: why do we need to probe? sync CLI will poll internally and send result so probing should not be here

File: dropin/index.js [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+// First-build detection for the sync classifier (KD7). With no committed baselines the head run
+// itself becomes the project's baseline: `percy exec` sends the baseline-candidate flag and the
+// API rewrites the build's source iff it is the project's first build. The CLI exposes the decided
+// source through the healthcheck build info — when this run's build IS the baseline, its diffs are
+// baseline-establishment noise. The reporter (Gate A) does the authoritative post-finish detection.
+function isFirstBuildRun() {
+  return Boolean(utils.percy && utils.percy.build &&
+    utils.percy.build.source === 'playwright-dropin-baseline');
+}
```
Comment (rishigupta1599) [inline dropin/index.js]: this is repeating multiple places, have one util function

File: commands/setup-baseline.mjs (hunk 1 above)
Comment (Shivanshu-07) [reply commands/setup-baseline.mjs]: Good catch — extracted to a shared helper in 6077ec4: `dropin/config.js` now exports `SUPPORTED_PROJECT_TYPES` + `wrongTokenError()`, used by both `validateConfig` and this command, so the check and wording live in one place.

File: dropin/baseline/config-reporter.js (hunk 2 above)
Comment (Shivanshu-07) [reply dropin/baseline/config-reporter.js:44]: Local-only handoff, never uploaded: `resolve-config.js` spawns `playwright test --list` with this reporter so Playwright itself resolves the config (TS/ESM included), reads this temp file immediately after the process exits, and deletes it in a `finally`. Added a clarifying comment at the write site in 6077ec4.

File: dropin/png.js (hunk 3 above)
Comment (Shivanshu-07) [reply dropin/png.js:15]: Deliberately dependency-free: we only need width/height from the PNG IHDR header (two u32 reads at fixed offsets, signature-checked, null-safe on anything malformed). Full PNG packages pull in decoders we don't use, and `image-size` cost the CLI a Node-support pin (percy/cli#2301). Happy to revisit if we ever need real PNG operations.

File: dropin/version-info.js (hunk 4 above)
Comment (Shivanshu-07) [reply dropin/version-info.js]: You're right — the root SDK module already computes and exports `CLIENT_INFO`/`ENV_INFO` (they're sent as clientInfo/environmentInfo on every post, and sdk-utils folds them into the request user agent). Removed the duplicate module in 6077ec4; the drop-in now reads them from the root, required lazily because the drop-in loads inside `playwright.config` before the runner finishes booting.

File: dropin/sync.js (hunk 5 above)
Comment (Shivanshu-07) [reply dropin/sync.js]: Agreed — removed in 6077ec4. With sync the CLI polls internally and returns the verdict on the post itself, so the SDK never needs a read of its own; a genuinely unauthorized token surfaces as a 403 in the classifier's no-verdict bucket, backstopped by the gate. `preflightTokenScope` and its validateConfig call are gone.

File: dropin/index.js (hunk 6 above)
Comment (Shivanshu-07) [reply dropin/index.js]: Consolidated in 6077ec4 — new `dropin/percy-info.js` owns the `utils.percy` null-chains and the baseline-source literal (`percyBuild`/`percyProjectType`/`isBaselineBuildRun`), now used by index.js and the reporter.

**Review by rishigupta1599 — DISMISSED** [review]

**Review by rishigupta1599 — APPROVED** [review]
